File: Discrete Mathematics/DataStructure TS/graph.ts
import fs from 'fs'
import { adjacentListI} from './graph.interface'
class Graph {
    numberOfNodes:number
    adjacentList: adjacentListI
    constructor() {
        this.numberOfNodes = 0;
        this.adjacentList = {
        };
    }
    
    addVertex(node:string): adjacentListI {
        this.adjacentList[node] = []
    
        this.numberOfNodes++;
        return this.adjacentList
    }
    addEdge(node1:string, node2:string) {
        //undirected Graph 
        this.adjacentList[node1].push(node2)
        this.adjacentList[node2].push(node1)
        return this.adjacentList[node1]
    }
    isPattern(node1:string) {
        const allNodes = Object.keys(this.adjacentList);
        let temp  = 0 ;
        for(let node of allNodes) {
            for (let char of node ) {
                if (node1.includes(char)) {
                    temp +=1;
                    if (temp == 4) {
                        return node;
                    }
                }
            }
        }
        return null;
    }
    showConnections() {
        const allNodes = Object.keys(this.adjacentList);
        for (let node of allNodes) {
            let nodeConnections = this.adjacentList[node];
            let connections = "";
            let vertex;
            for (vertex of nodeConnections) {
                connections += vertex + " ";
            }
            console.log(node + "-->" + connections);
        }
        console.log(this.adjacentList)
    }
}
// const wordsList:string[] = []


// Init Graph
const myGraph = new Graph();
const list_words: string[] = fs.readFileSync('../sgb-words-test.txt', 'utf-8').split('\n')

// Add Vertex 

for (let word of list_words) {
   
    const nodePattern = myGraph.isPattern(word)
    myGraph.addVertex(word);
    if (nodePattern) {
        myGraph.addEdge(word,nodePattern)
    }
}

// myGraph.addVertex('0');
// myGraph.addVertex('1');
// myGraph.addVertex('2');
// myGraph.addVertex('3');
// myGraph.addVertex('4');
// myGraph.addVertex('5');
// myGraph.addVertex('6');
// myGraph.addEdge('3', '1');
// myGraph.addEdge('3', '4');
// myGraph.addEdge('4', '2');
// myGraph.addEdge('4', '5');
// myGraph.addEdge('1', '2');
// myGraph.addEdge('1', '0');
// myGraph.addEdge('0', '2');
// myGraph.addEdge('6', '5');

myGraph.showConnections();

File: Discrete Mathematics/Graph.py

class GraphUndirected():
    def __init__(self):
        self.numberOfNodes = 0 
        self.adjacentList = {}
        self.vertex_visited = {}
    
    def addVertex(self,node):
        self.adjacentList[node] = []
        self.numberOfNodes+=1
        return self.adjacentList
    
    def addEdge(self,node1, node2):
        ## undirected Graph
        self.adjacentList[node1].append(node2)
        self.adjacentList[node2].append(node1)
        return self.adjacentList[node1]
    
    def showConnections(self):
        allNodes = self.get_all_nodes()
        for node in allNodes:
            nodeConnections = self.adjacentList[node]
            connections = ""
            for vertex in nodeConnections:
                connections += vertex + " "
            print(node + "-->" + connections)
        
        # print(self.adjacentList)
    def is_the_sample_unique_character(self,word1,word2):
        temp = 0
        for cha1 in word1:
            if cha1 in word2:
                temp +=1
                word2 = word2.replace(cha1,'',1)
                if temp == 4:
                    return True
        return False

    def isPattern(self,node1):
        allNodes = self.get_all_nodes()

        list_words = []
        for node in allNodes:
            if self.is_the_sample_unique_character(node,node1):
                list_words.append(node)
        return list_words
    
    def init_visited_vertex(self):
        allNodes = self.get_all_nodes()
        for node in allNodes:
            self.vertex_visited[node] = False
        return self.vertex_visited
        
    def dfs(self):
        cc = 0
        allNodes = self.get_all_nodes()
        vertex_visited = self.init_visited_vertex()
        for node in allNodes:
            if not vertex_visited[node]:
                cc +=1
                self.expolore(node,self.adjacentList[node])
        print(cc)
    def expolore(self,node,graph):
        self.vertex_visited[node] = True
        for edge in graph:
            if not self.vertex_visited[edge]:
                self.expolore(edge, self.adjacentList[edge])
        
    def get_all_nodes(self):
        return list(self.adjacentList.keys())

    def find_path(self, start_vertex, end_vertex, path=None):
        """ 
        find a path from start_vertex to end_vertex 
            in graph 
        """
        if path == None:
            path = []
        graph = self.adjacentList
        path = path + [start_vertex]
        if start_vertex == end_vertex:
            return path
        if start_vertex not in graph:
            return None
        for vertex in graph[start_vertex]:
            if vertex not in path:
                extended_path = self.find_path(vertex,
                                               end_vertex,
                                               path)
                if extended_path:
                    return extended_path
        return None
